### PR TITLE
Group shape/dtype validation logic in image_resample.

### PR DESCRIPTION
**Pull request by anntzer**
Move it all to a single place rather than having some of it interspersed with the dtype dispatch later.

Also reorder the dtype dispatch to be consistent in the 2D and 3D cases, and remove _array from many local variable names.

## PR Checklist

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Review by greglucas — APPROVED** [review]
Looks good to me. Ideally, this would get some `pytest.raises(ValueError, "match")` for the updated error strings that are missing coverage.

**Comment by anntzer** [comment]
Sure, added tests.

File: src/_image_wrapper.cpp [hunk 1]
```diff
@@ -143,36 +144,52 @@ image_resample(PyObject *self, PyObject* args, PyObject *kwargs)
...
-            &py_input_array, &py_output_array, &py_transform,
+            &py_input, &py_output, &py_transform,
             &params.interpolation, &convert_bool, &params.resample,
             &params.alpha, &convert_bool, &params.norm, &params.radius)) {
         return NULL;
     }
 
     if (params.interpolation < 0 || params.interpolation >= _n_interpolation) {
-        PyErr_Format(PyExc_ValueError, "invalid interpolation value %d",
+        PyErr_Format(PyExc_ValueError, "INvalid interpolation value %d",
```
Comment (oscargus) [inline src/_image_wrapper.cpp:154]:
```suggestion
        PyErr_Format(PyExc_ValueError, "Invalid interpolation value %d",
```
(or was this on purpose?)

**Review by oscargus — APPROVED** [review]
Minor capitalization thingy, but otherwise looks good!

File: src/_image_wrapper.cpp (hunk 1 above)
Comment (anntzer) [reply src/_image_wrapper.cpp:154]: oops, fixed.

**Pull request merged**